src/Http/Http/src/StreamPipeReader.cs

```diff
@@ -0,0 +1,345 @@
+                returnStart = returnStart.NextSegment;
+            }
+        }
+
+        /// <inheritdoc />
+        public override void CancelPendingRead()
+        {
+            InternalTokenSource.Cancel();
+        }
+
+        /// <inheritdoc />
+        public override void Complete(Exception exception = null)
+        {
+            if (_isCompleted)
+            {
+                return;
+            }
+
+            _isCompleted = true;
+            if (exception != null)
+            {
+                _exceptionInfo = ExceptionDispatchInfo.Capture(exception);
+            }
+
+            var segment = _readHead;
+            while (segment != null)
+            {
+                segment.ResetMemory();
+                segment = segment.NextSegment;
+            }
+        }
+
+        /// <inheritdoc />
+        public override void OnWriterCompleted(Action<Exception, object> callback, object state)
+        {
+            throw new NotSupportedException("OnWriterCompleted is not supported");
+        }
+
+        /// <inheritdoc />
+        public override async ValueTask<ReadResult> ReadAsync(CancellationToken cancellationToken = default)
+        {
+            // TODO ReadyAsync needs to throw if there are overlapping reads.
+            ThrowIfCompleted();
+
+            // PERF: store InternalTokenSource locally to avoid querying it twice (which acquires a lock)
+            var tokenSource = InternalTokenSource;
+            if (TryReadInternal(tokenSource, out var readResult))
+            {
+                return readResult;
+            }
+
+            var reg = new CancellationTokenRegistration();
+            if (cancellationToken.CanBeCanceled)
+            {
+                reg = cancellationToken.Register(state => ((StreamPipeReader)state).Cancel(), this);
+            }
+
+            using (reg)
+            {
+                var isCanceled = false;
+                try
+                {
+                    AllocateReadTail();
+#if NETCOREAPP2_2
+                    var length = await _readingStream.ReadAsync(_readTail.AvailableMemory.Slice(_readTail.End), tokenSource.Token);
+#elif NETSTANDARD2_0
+                    if (!MemoryMarshal.TryGetArray<byte>(_readTail.AvailableMemory.Slice(_readTail.End), out var arraySegment))
+                    {
+                        ThrowHelper.CreateInvalidOperationException_NoArrayFromMemory();
+                    }
+
+                    var length = await _readingStream.ReadAsync(arraySegment.Array, arraySegment.Offset, arraySegment.Count, tokenSource.Token);
+#else
+#error Target frameworks need to be updated.
+#endif
+                    Debug.Assert(length + _readTail.End <= _readTail.AvailableMemory.Length);
+
+                    _readTail.End += length;
+                    _bufferedBytes += length;
+                }
+                catch (OperationCanceledException)
+                {
+                    ClearOutCancellation();
```

Comment:
When a read was canceled by the user token instead of a call to CancelPendingRead() meaning ReadAsync throws instead of returning a CanceledReadResult,  a real PipeReader would not clear or observe the cancelation.

Reply:
“real” PipeReader

Reply:
We have to clear the cancellation because when the user token is canceled, we always cancel the _internalTokenSource. We pass in the _internalTokenSource.Token to ReadAsync and register a callback on the user token to cancel the _internalTokenSource.

A potential interleaving you may be worried about is the user token being tripped and then someone calling CancelPendingRead() immediately afterwards. If CancelPendingRead is called before ClearOutCancellation(), it would mean the next call to ReadAsync would work (and not be canceled). If CancelPendingRead is called afterwards, it means the next call to ReadAsync will be canceled immediately. Both described behaviors are "fine," as there is no guarantee that calling CancelPendingFlush() will after tripping the user cts will cause the current or next read to fail.

Reply:
>  there is no guarantee that calling CancelPendingFlush() will after tripping the user cts will cause the current or next read to fail.

The distinction is in *how* the current or next read fails. Cancellations caused by CancellationTokens are treated very differently than cancelations caused by CancelPendingRead/Flush().

With the "real" PipeReader, even if the a ReadAsync() invocation throws an OperationCanceledException due to a tripped token after CancelPendingRead() was called, the next ReadAsync() call without a tripped token will still return a canceled ReadResult.

That being said, it's unusual to continue reading after an OperationCanceledException, so feel free to ignore this until a "real" person complains.
differenceCancelPendingRead/Flush is guaranteed to cause set IsCanceled